### DISCUSSION TITLE
How to increase the number of object references in C?

### DISCUSSION THREAD

**Discussion by Mars-CN** [micropython/micropython#13397] [General]
When using C to extend MicroPython, some objects are created in the Python environment and directly passed to C through functions for processing. However, in the Python environment, variables are not assigned, which may result in garbage collection during GC. When using C to extend Python, macros such as "Py-INCREF" and "Py-DECREF" can be called to define the number of references to the operation object. In MicroPython development, I haven't found any similar macro definitions or functions, how should I solve them?

**Comment by dlech** [comment]
MicroPython doesn't use reference counting. Instead, if you need to keep an object alive, it needs to be pointed to via an object reachable through a "root pointer". So you can make a pointer from some other MicroPython object to the one that gets passed to C. Or you can create your own root pointers if needed with the `MP_REGISTER_ROOT_POINTER` macro. But since you didn't give much detail on your specific case, it's hard to say what would be the best way to handle it.

**Reply by Mars-CN** [reply]
For example, in the Python environment, there is a piece of code like this
import mod
brush = mod.Brush()
brush.font(mod.Font("/font.ttf"))

At this point, the object generated by mod.Font("/font.ttf") does not generate a reference in the Python environment. If the test calls gc.collect(), this anonymous object will be reclaimed, which is an undesirable result.

So we should add a reference to the font function, which is an extension of C, how can we add a reference to this object in C?

**Comment by Mars-CN** [comment]
For example, in the Python environment, there is a piece of code like this
```
import mod
brush = mod.Brush()
brush.font(mod.Font("/font.ttf"))
```

At this point, the object generated by `mod.Font("/font.ttf")` does not generate a reference in the Python environment. If the test calls gc.collect(), this anonymous object will be reclaimed, which is an undesirable result.

So we should add a reference to the font function, which is an extension of C, how can we add a reference to this object in C?

**Reply by dlech** [reply]
Since `brush` is a Python object, I would add a `mp_obj_t font` member to the object instance struct to hold the reference. Then in the `font()` method implementation, assign the argument to that new field and keep it there as long a the C code is using it.

**Reply by Mars-CN** [reply]
However, this solution will cause the Font variable reference to be lost after calling `gc.collect()`

**Reply by xyzzy42** [reply]
That shouldn't happen.  The `bush` object needs to keep a reference to the Font object that is passed to it in the font() method call.  That will it won't get GCed until the brush itself is GCed.

**Reply by Mars-CN** [reply]
But in reality, it happened. It can be understood that in the Python environment, a 'Font' variable defined through 'brush. font (mod. Font ("/font. ttf")' has its declaration cycle management handed over to the Python VM. If 'brush. font (mod. Font ("/font. ttf") "is used again, the previously defined variable should be reclaimed by GC, which is expected. However, in reality, after calling ` brush. font (mod. Font ("/font. ttf") `, the ` mod. " Font ("/font. ttf") ` has already lost its reference in Python's MV and will be directly reclaimed by GC.
There is another situation where if the 'Font' variable defined through 'brush. font (mod. Font ("/font. ttf")' has jurisdiction over the C environment, then when calling 'brush. font (mod. Font ("/font2. ttf") "again, the previous' Font' will not be automatically destroyed. If not manually destroyed, it will cause a memory leak. However, when manually destroyed, the C environment does not know if it has a reference in the Python VM, which can lead to a memory overflow error.

```
Font=mod Font ("/font. ttf")
Brush. font (font)
Font2=mod Font ("/font2. ttf")
Brush. font (font2)
```
In the above code, both font and font2 are referenced, so they will not be reclaimed by GC.
however
```
Font=mod Font ("/font. ttf")
Brush. font (font)
Font=mod Font ("/font2. ttf")
Brush. font (font2)
```
In the third line of code, the font defined for the first time has lost its reference and will be reclaimed by GC.
So:

```
Brush. font (mod. Font ("/font. ttf"))
Font=mod Font ("/font2. ttf")
Brush. font (font2)
```
How does VM handle this situation?

**Reply by glenn20** [reply]
The code examples here appear to have been mangled. It will help to make sure there are no ambiguities if you could edit the post to fix the code blocks.

Nonetheless, I think I understand what you are asking.

```py
import mod
brush = mod.Brush()
font = mod.Font("/font1.ttf")
brush.font(font)
font = mod.Font("/font2.ttf")
brush.font(font2)
```

If you implement the suggestion by @dlech to add a member to the `Brush` object which holds a refeence to the `Font` object, then every `Brush` object will hold a reference to the last `Font` object it is using, so:

- After the fourth line both the `font` and `brush` objects hold a reference to the `Font` object for "font1.ttf" (preventing gc collection).
- After the fifth line only the `brush` object holds a reference to the `Font` object for "font1.ttf" (preventing gc collection).
- After the sixth line no objects hold a reference to the `Font` object for "font1.ttf" and it may be collected in the next `gc.collect()`.
